### PR TITLE
feat: Add NPU+GPU async pipelining for vision-language models

### PR DESCRIPTION
**Pull request by liangliangchang**
## Purpose

Enable **NPU+GPU async pipelining** for vision-language models to improve concurrent request throughput on heterogeneous hardware (NPU for vision encoding + GPU for LLM inference).

**Problem**: When using NPU for vision encoding and GPU for LLM, requests process sequentially even when submitted concurrently, because the V1 scheduler cannot pipeline different stages of different requests when `max-num-seqs=1`:
```
- Request 1: [NPU Vision 13s] →[GPU LLM 30s]
- Request 2:                                     [NPU Vision 13s] →[GPU LLM 30s]
```
- **Total: ~130s for 3 requests**

**Solution**: This PR adds a **Vision Scheduler** that proactively processes NPU vision encoding for waiting requests in background threads, enabling true hardware parallelism:
```
- Request 1: [NPU Vision 13s] → [GPU LLM 30s]
- Request 2:                    [NPU Vision 13s] → [GPU LLM 30s]← Overlaps with Req1 decode!
- Request 3:                                       [NPU Vision 13s] → [GPU LLM 30s]
```
- **Total: ~103s for 3 requests**

### Key Components

1. **Vision Scheduler** (`vllm/v1/engine/core.py`):
   - Runs before core scheduler in every `step()`
   - Scans all WAITING requests with vision features
   - Proactively submits vision encoding to background thread pool
   - Works independently of `max_num_running_reqs` setting

2. **Background Vision Pre-encoding** (`vllm/v1/worker/gpu_model_runner.py`):
   - Thread pool for NPU vision processing
   - **GIL release** during C++ FlexMLRT inference (enables true parallelism)
   - Vision embeddings cached for instant scheduling

3. **Hybrid Scheduler Check** (`vllm/v1/core/sched/scheduler.py`):
   - Checks vision readiness before scheduling WAITING requests
   - Defers requests whose vision isn't ready yet

### Environment Variables

Enable with:
```bash
export VLLM_NPU_ASYNC_PIPELINE=1
export VLLM_VISION_NPU_BACKEND=flexmlrt 
export VLLM_VISION_NPU_DEVICE=stx
export VLLM_VISION_NPU_CACHE=/path/to/npu_model cache
```

Gracefully degrades when disabled (zero overhead for GPU-only mode).

### Prerequisites
- AMD Strix NPU hardware (or compatible NPU)
- FlexMLRT runtime installed
- Qwen2.5-VL model

### Test 1: NPU Mode Performance

```bash
cd tests/async_pipelining

# Start NPU server
./start_vllm_server.sh

# Run NPU performance test (in another terminal)
python compare_npu_vs_gpu.py --mode npu
```

**Expected behavior**:
- Server logs show "Deferred: N reqs" (vision processing in background)
- Concurrent speedup vs sequential baseline (~1.4×)
- Vision encoding happens proactively for waiting requests

### Test 2: GPU-only Mode (Baseline)

```bash
# Stop NPU server
pkill -f vllm.entrypoints.openai.api_server

# Start GPU-only server
./test_pure_gpu.sh

# Run GPU performance test
python compare_npu_vs_gpu.py --mode gpu --num-requests 3
```

**Expected behavior**:
- No "Deferred" requests (no Vision Scheduler active)
- Concurrent speedup from reduced cold-start overhead

### Test 3: Compare Results

```bash
python compare_npu_vs_gpu.py --compare
```

### Test 4: End-to-end Pipelining Validation

```bash
# With NPU server running
python test_server_async_pipelining.py
```

**Verify in logs**:
- Vision encoding timestamps overlap with LLM decode
- "Deferred" count increases when vision not ready
- Performance improvement over sequential baseline

### Hardware Setup
- **Platform**: AMD Krackan 2E (integrated GPU) + Strix NPU
- **Model**: Qwen2.5-VL-7B-Instruct (W4A16 quantized)
- **Config**: `--max-num-seqs 1 --gpu-memory-utilization 0.83`
- **NPU vision time**: ~8.6s per image
- **GPU decode time**: ~30s per request (100 tokens @ 2.0 tok/s)

#### Timing breakdown
| Platform | Device | Vision | Prefill | Decode | Peak tok/s |
|----------|--------|--------|---------|--------|------------|
| Strix | NPU | 7.688s | 3.082s | 11.523s | 12 |
| Strix | GPU |2.105s |3.082s  | 11.523s  |12  |
| K2e | NPU | 8.913s | 19.558s | 35.248s | 2.2 |
| K2e | GPU | 9.582s | 19.558s | 35.248s |2.2  |
#### End-to-end time (3 concurrent requests)
| Platform | Configuration | Time |
|----------|---------------|------|
| Strix | NPU+GPU | 49.27s |
| Strix | GPU-only | 47.14s |
| K2e | NPU+GPU | 175.92s |
| K2e | GPU-only | 192.58s |

The "Deferred" requests prove the Vision Scheduler is working - requests wait for vision to complete while not blocking the scheduler.

### Output Quality Verification

Sample outputs from both modes (NPU vs GPU) are semantically identical, confirming NPU vision encoding produces correct results:

**NPU+GPU Output**:
```
The image depicts a stunning waterfall cascading down a rocky cliff...
```

**Pure GPU Output**:
```
The image depicts a stunning waterfall cascading down a rocky cliff...
```

### Files Changed

```
27 files changed, 3,229 insertions(+), 18 deletions(-)
```

**Core pipelining** (939 lines):
- `vllm/v1/engine/core.py` (+112 lines)
- `vllm/v1/worker/gpu_model_runner.py` (+789 lines)
- `vllm/v1/core/sched/scheduler.py` (+41 lines)
- `vllm/v1/engine/output_processor.py` (+24 lines)
- `vllm/v1/executor/uniproc_executor.py` (+20 lines)

**NPU backend** (1,193 lines):
- `vllm/vision_npu/*`: FlexMLRT backend, CPU preprocessing, C++ bridge

**Test suite** (1,097 lines):
- `tests/async_pipelining/*`: Performance comparison tools, test scripts, test images

---

## Essential Elements Checklist

- [x] **Purpose**: Enable NPU+GPU async pipelining for 1.4× throughput improvement on heterogeneous hardware
- [x] **Test plan**: Provided test commands for NPU mode, GPU-only mode, and comparison
- [x] **Test results**: Performance numbers showing 1.26× speedup (192s vs 241s for 3 concurrent requests)
- [x] **Documentation**: Added comprehensive README in `tests/async_pipelining/README.md`
- [ ] **Release notes**: N/A - This is a hardware-specific optimization for NPU users, not a general user-facing feature

### Additional Notes

- **Zero overhead when disabled**: All changes are gated by environment variables
- **Backward compatible**: Works with existing V1 async scheduler
- **Graceful degradation**: Falls back to standard behavior when NPU unavailable
- **No changes to GPU-only path**: GPU-only mode unchanged and unaffected

### CODE REVIEW

File: vllm/vision_npu/cpu_preprocess.py [hunk 1]
```diff
@@ -0,0 +1,279 @@
...
+
+import logging
+
+import numpy as np
+import torch
+
+logger = logging.getLogger(__name__)
+
+
+class Qwen2_5_VL_CPUPreprocessor:
```
Comment (mgehre-amd) [inline vllm/vision_npu/cpu_preprocess.py:24]: Seems strange that a model-specific class in in the generic vllm/vision_npu package. Can this be split better?

File: vllm/vision_npu/bridge/CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1,45 @@
+# SPDX-License-Identifier: Apache-2.0
```
Comment (mgehre-amd) [inline vllm/vision_npu/bridge/CMakeLists.txt:1]: Please include reproducible build instructions in the PR description and modify the README.md accordingly.
Make sure to only use the public release of Ryzen AI.

File: vllm/envs.py [hunk 3]
```diff
@@ -1744,6 +1747,12 @@ def _get_or_set_default() -> str:
     # Disable PDL for LoRA, as enabling PDL with LoRA on SM100 causes
     # Triton compilation to fail.
     "VLLM_LORA_DISABLE_PDL": lambda: bool(int(os.getenv("VLLM_LORA_DISABLE_PDL", "0"))),
+    # NPU vision backend to use (e.g., "flexmlrt" for FlexMLRT backend)
+    "VLLM_VISION_NPU_BACKEND": lambda: os.getenv("VLLM_VISION_NPU_BACKEND", ""),
```
Comment (mgehre-amd) [inline vllm/envs.py:1753]: There is only one backend right? I.e. we got drop this env var?

File: vllm/model_executor/models/qwen2_5_vl.py [hunk 4]
```diff
@@ -581,18 +581,40 @@ def __init__(
     ) -> None:
         super().__init__()
 
+        # Store minimal config needed for both NPU and PyTorch paths
```
Comment (mgehre-amd) [inline vllm/model_executor/models/qwen2_5_vl.py:584]: I wonder whether making a new Qwen2_5_VisionTransformerNPU class would be cleaner instead of doing the overwriting/conditionals here. Can you try that?

**Comment by mgehre-amd** [comment]
Hi, thanks for the PR!
Can you please clarify what you mean by "cold start overhead"? Where did you find that?

I think you performance comparisons are not fair. We use "--max-num-seqs 1" to measure single request latency.

If you want to measure throughput under the assumption that two requests will be available, you should use "--max-num-seqs 2" for GPU-only side.

**Comment by liangliangchang** [comment]
> Hi, thanks for the PR! Can you please clarify what you mean by "cold start overhead"? Where did you find that?
> 
> I think you performance comparisons are not fair. We use "--max-num-seqs 1" to measure single request latency.
> 
> If you want to measure throughput under the assumption that two requests will be available, you should use "--max-num-seqs 2" for GPU-only side.

In vLLM, if we set max-num-seqs to be from 1 to 3, its speed is almost the same. Because vLLM will process up to 3 requests together in one iteration using SPMD-style in GPU. The current NPU vision model we extracted can only process 1 request. So it is not a fair comparison in that case. That's the reason max-seq-num is set to 1 during the test. But 3 requests were also sent in parallel for the GPU test, and it performed better than sending sequentially. It seems some startup, pre/post processing overhead gets avoided.

File: vllm/vision_npu/bridge/CMakeLists.txt (hunk 2 above)
Comment (liangliangchang) [reply vllm/vision_npu/bridge/CMakeLists.txt:1]: Yes, I also want to ask, what is the best method we can use to handle the NPU model cache? We have two options: 1. providing scripts and instructions, so users can generate their own NPU model cache by extracting the vision model from Hugging Face VL models and compile it using release Ryzen AI toolflow. 2. just providing the model cache so users don't have to worry about model extraction and compilation. Which one looks better to you? In both cases, we will provide instructions on how to install flexmlRT from the release Ryzen AI pkgs.

File: vllm/compilation/passes/fusion/act_quant_fusion.py [hunk 5]
```diff
@@ -39,7 +39,10 @@
 if silu_and_mul_nvfp4_quant_supported:
     FUSED_OPS[kNvfp4Dynamic] = torch.ops._C.silu_and_mul_nvfp4_quant.default  # noqa: E501
 
-if current_platform.is_cuda_alike():
+# Check if the per-block quant operation is available (newer ROCm/CUDA versions)
+if current_platform.is_cuda_alike() and hasattr(
+    torch.ops._C, "silu_and_mul_per_block_quant"
```
Comment (mgehre-amd) [inline vllm/compilation/passes/fusion/act_quant_fusion.py:44]: How is this related to the NPU work?

**Comment by mgehre-amd** [comment]
Can you check whether the NPU prefill can be integrated via the existing pipeline parallelism in vLLM (see `pipeline_parallel_size` option)? The current PR does pretty intrusive changes to core vLLM data structures,
which will make it hard for us to keep in sync with upstream vLLM (maintenance) and won't allow us to upstream.

**Comment by liangliangchang** [comment]
> Can you check whether the NPU prefill can be integrated via the existing pipeline parallelism in vLLM (see `pipeline_parallel_size` option)? The current PR does pretty intrusive changes to core vLLM data structures, which will make it hard for us to keep in sync with upstream vLLM (maintenance) and won't allow us to upstream.

Thanks. I agree. I only quickly checked the current pipeline parallelism during my implementation, and it seems to be for splitting fine-grained layers across GPUs. I will evaluate it in detail and see how to use the existing flow as much as possible. Will update later.
